openhands/llm/llm.py

```diff
@@ -491,19 +499,43 @@ def _post_completion(self, response: ModelResponse) -> float:
 
         return cur_cost
 
-    def get_token_count(self, messages) -> int:
-        """Get the number of tokens in a list of messages.
+    def get_token_count(self, messages: list[dict] | list[Message]) -> int:
```

Comment:
I think it's nice to add a custom tokenizer option! But one question: do you know why we have two different formats for messages? This seems like a problem that we should fix further upstream in the codebase, which would then allow us to simplify this function and presumably other places too.

Maybe we could open an issue for this?

Reply:
Ah, thank you. Maybe we shouldn't. I'll think about this and address it in a follow-up.

Reply:
The question that lead to this solution - is what exactly does our code need, core code or eval code, to get more accurate token counts for. Do we need, or find it easier to use for ... which of those two?

- litellm has a possibility to return token counts from some providers, which is addressed here https://github.com/All-Hands-AI/OpenHands/pull/5604
- so we need Messages, to save those counts
- Message is our class, not litellm's Message class, a part of the reasoning is [here](https://github.com/All-Hands-AI/OpenHands/blob/da71e73c4a34b74c202738eb38c69523ba0a642a/openhands/core/message_format.md)
- the litellm token counting [needs dicts](https://github.com/BerriAI/litellm/blob/fd583e715ec1b183cade08ac29c1cbe73e5c94f6/litellm/utils.py#L1517).

So we kinda need both. That seems rather... understandable? It seems reasonable for us to use our Message class for most things we need to do with messages (it was originally created to support vision models). The dicts here are really only its serialized form: the form sent to the litellm `completion` call (and to litellm's token counting :)

Maybe we can remove the dicts... idk, that's the native litellm form.

On another hand, what exactly do we need for token counts, has a third answer, and this one is amusing: we may also need token counts for Events, that is, for the dicts they will eventually become when sent to the LLM. That sounds strange - and I'm not suggesting to add that here! - but I'm working on it or related to it, because there are two apparent needs at the moment:
- PRs that attempt to use the context truncation we have, to implement truncation for 8k, 32k, etc. They write `llm.get_token_count(events)`. Developers seem to feel that that should work; that such feature belongs where the truncation is (in the controller), and there are only Events there. No Messages nor messages. I think this is actually not quite right, it doesn't belong in the controller, but then I think [our truncation](https://github.com/All-Hands-AI/OpenHands/blob/590ebb6e4730301b5644fb92b894ea391f3025c2/openhands/controller/agent_controller.py#L819) shouldn't really be there either. They both belong in the memory component IMO (which we had two attempts at, so far, and we removed each; third time will be a charm! 😂 - for example Calvin is [refactoring it here](https://github.com/All-Hands-AI/OpenHands/pull/5306/files#diff-736089350cac5eabbeeb848e5d3f1c4477e76d0128b2ba480adf721832d44b63R85))
- getting token counts for our historical data: that requires converting from Events to Messages outside the agent loop, and it's not entirely trivial, it looks [something like this](https://github.com/enyst/playground/pull/25/files#diff-ce633fd8c26acf714bb00706a0089a6dc9a5977c7b3c0c54f1d584e4fdade434R53) currently. I'm working on that because I need the token counts for our trajectories in .jsonl files, and they're saved as events. I understand we also save the messages dicts ([here](https://github.com/All-Hands-AI/OpenHands/blob/590ebb6e4730301b5644fb92b894ea391f3025c2/openhands/llm/llm.py#L262)), but they have not been made public. So I'm reconstructing them.

Pending the question whether the memory component should or could handle such event->message conversion thus could support the third, I think the latter may belong in a script in evals perhaps, rather than in core. Idk. I mention it for the sake of completeness.

Reply:
We could remove the dicts as parameter here, or at least remove them almost everywhere in the code. 🤔

We only need them as the form in which we send the data over to `litellm completion` or `token_counter`, otherwise _maybe_ we can work without them. That might pose a bit of a challenge when it comes to the non-native to native function calling [conversion](https://github.com/All-Hands-AI/OpenHands/blob/590ebb6e4730301b5644fb92b894ea391f3025c2/openhands/llm/fn_call_converter.py#L308) (it uses dicts everywhere), but I'll think about that.

- https://github.com/All-Hands-AI/OpenHands/issues/5616